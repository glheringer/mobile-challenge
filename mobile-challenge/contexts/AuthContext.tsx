import React, { createContext, useState, useEffect, ReactNode } from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";
import axios from "axios";

interface AuthContextProps {
  user: any;
  signIn: (username: string, password: string) => Promise<boolean>;
  signOut: () => Promise<void>;
}

const AuthContext = createContext<AuthContextProps | undefined>(undefined);

interface AuthProviderProps {
  children: ReactNode;
}

const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
  const [user, setUser] = useState(null);

  const signIn = async (
    username: string,
    password: string
  ): Promise<boolean> => {
    try {
      const response = await axios.post(
        "https://test-api-y04b.onrender.com/signIn",
        { user: username, password }
      );

      console.log("Response data:", response.data); // Adicione este log para depuração

      if (!response.data.error) {
        setUser(response.data.user);
        await AsyncStorage.setItem("user", JSON.stringify(response.data.user));
        return true;
      } else {
        return false;
      }
    } catch (error) {
      console.error("Erro ao fazer login:", error);
      return false;
    }
  };

  const signOut = async () => {
    try {
      await AsyncStorage.removeItem("user");
      setUser(null);
    } catch (error) {
      console.error("Erro ao fazer logout:", error);
    }
  };

  useEffect(() => {
    const loadUser = async () => {
      const storedUser = await AsyncStorage.getItem("user");
      if (storedUser) {
        setUser(JSON.parse(storedUser));
      }
    };

    loadUser();
  }, []);

  return (
    <AuthContext.Provider value={{ user, signIn, signOut }}>
      {children}
    </AuthContext.Provider>
  );
};

export { AuthContext, AuthProvider };
