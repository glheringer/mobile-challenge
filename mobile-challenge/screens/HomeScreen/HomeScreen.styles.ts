import styled from "styled-components/native";
import { theme } from "utils/theme";
import HeaderLogoComponent from "components/HeaderLogo/Header";

export const Container = styled.View`
  flex: 1;
  padding: ${theme.spacings.medium};
  background-color: ${theme.colors.background};
`;

export const UserName = styled.Text`
  font-size: ${theme.fontSizes.small};
  font-family: ${theme.fonts.bold};
  color: ${theme.colors.primary};
  text-align: center;
  margin-bottom: ${theme.spacings.medium};
`;

export const BrandContainer = styled.TouchableOpacity`
  flex: 1;
  margin: ${theme.spacings.small};
  padding: ${theme.spacings.medium};
  background-color: ${theme.colors.primary};
  border-width: 1px;
  border-color: ${theme.colors.secondary};
  border-radius: 8px;
  align-items: center;
  justify-content: center;
`;

export const BrandText = styled.Text`
  font-size: ${theme.fontSizes.medium};
  font-family: ${theme.fonts.regular};
  color: ${theme.colors.buttonText};
`;

export const EmptyText = styled.Text`
  font-size: ${theme.fontSizes.medium};
  color: ${theme.colors.secondaryText};
  text-align: center;
  margin-top: ${theme.spacings.medium};
`;
