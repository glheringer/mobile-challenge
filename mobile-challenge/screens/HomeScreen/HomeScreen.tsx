import React, { useContext, useEffect, useState } from "react";
import { Button, FlatList } from "react-native";
import axios from "axios";
import { AuthContext } from "contexts/AuthContext";
import styled from "styled-components/native";
import { theme } from "utils/theme";
import { logo } from "../../assets";

import { AppStackParamList } from "App";
import { StackNavigationProp } from "@react-navigation/stack";
import { useNavigation } from "@react-navigation/native";

import Header from "components/HeaderLogo/Header";

type HomeScreenNavigationProp = StackNavigationProp<
  AppStackParamList,
  "Models"
>;

interface Brand {
  codigo: string;
  nome: string;
}

const HomeScreen: React.FC = () => {
  const [brands, setBrands] = useState<Brand[]>([]);
  const navigation = useNavigation<HomeScreenNavigationProp>();
  const authContext = useContext(AuthContext);

  useEffect(() => {
    const fetchBrands = async () => {
      const response = await axios.get(
        "https://parallelum.com.br/fipe/api/v1/carros/marcas"
      );
      setBrands(response.data);
    };

    fetchBrands();
  }, []);

  const handlePress = (brandCode: string) => {
    navigation.navigate("Models", { brandCode });
  };

  const handleSignOut = () => {
    authContext?.signOut();
    navigation.navigate("Login");
  };

  return (
    <Container>
      <Header
        logo={logo}
        onSignOut={handleSignOut}
        title={"App de Marcas"}
        buttonColor={theme.colors.primary}
      />
      <UserName>Bem vindo, {authContext?.user?.name}</UserName>
      <FlatList
        data={brands}
        keyExtractor={(item) => item.codigo}
        renderItem={({ item }) => (
          <BrandContainer onPress={() => handlePress(item.codigo)}>
            <BrandText>{item.nome}</BrandText>
          </BrandContainer>
        )}
        numColumns={2}
        columnWrapperStyle={{ justifyContent: "space-between" }}
        ListEmptyComponent={<EmptyText>No brands found</EmptyText>}
      />
    </Container>
  );
};

const Container = styled.View`
  flex: 1;
  padding: ${theme.spacings.medium};
  background-color: ${theme.colors.background};
`;

const UserName = styled.Text`
  font-size: ${theme.fontSizes.small};
  font-family: ${theme.fonts.bold};
  color: ${theme.colors.primary};
  text-align: center;
  margin-bottom: ${theme.spacings.medium};
`;

const BrandContainer = styled.TouchableOpacity`
  flex: 1;
  margin: ${theme.spacings.small};
  padding: ${theme.spacings.medium};
  background-color: ${theme.colors.primary};
  border-width: 1px;
  border-color: ${theme.colors.secondary};
  border-radius: 8px;
  align-items: center;
  justify-content: center;
`;

const BrandText = styled.Text`
  font-size: ${theme.fontSizes.medium};
  font-family: ${theme.fonts.regular};
  color: ${theme.colors.buttonText};
`;

const EmptyText = styled.Text`
  font-size: ${theme.fontSizes.medium};
  color: ${theme.colors.secondaryText};
  text-align: center;
  margin-top: ${theme.spacings.medium};
`;

export default HomeScreen;
