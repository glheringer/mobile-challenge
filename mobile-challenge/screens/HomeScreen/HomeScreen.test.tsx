import React from "react";
import { render } from "@testing-library/react-native";
import HomeScreen from "./HomeScreen";
import axios from "axios";
import { AuthContext } from "../../contexts/AuthContext";
import { NavigationContainer } from "@react-navigation/native";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

const mockAuthContext = {
  user: { name: "Test User" },
  signIn: jest.fn(),
  signOut: jest.fn(),
};

describe("HomeScreen", () => {
  beforeEach(() => {
    mockedAxios.get.mockResolvedValue({ data: [] });
  });

  it("renders correctly and matches snapshot", async () => {
    const { toJSON } = render(
      <NavigationContainer>
        <AuthContext.Provider value={mockAuthContext}>
          <HomeScreen />
        </AuthContext.Provider>
      </NavigationContainer>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
