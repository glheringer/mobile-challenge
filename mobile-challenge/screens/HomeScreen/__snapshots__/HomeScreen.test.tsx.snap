// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly and matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
    }
  }
>
  <RCTSafeAreaView
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 16,
        "paddingBottom": 0,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 0,
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/logo.png",
        }
      }
      style={
        {
          "borderBottomLeftRadius": "50%",
          "borderBottomRightRadius": "50%",
          "borderTopLeftRadius": "50%",
          "borderTopRightRadius": "50%",
          "height": 60,
          "width": 60,
        }
      }
    />
    <View
      style={
        {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#145DA0",
            "fontSize": 24,
          }
        }
      >
        App de Marcas
      </Text>
    </View>
    <View
      style={
        {
          "overflow": "hidden",
        }
      }
    >
      <View
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#EF5350",
                  "fontSize": 40,
                },
                {
                  "backgroundColor": "transparent",
                },
                {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
            testID="iconIcon"
          >
            Óóç
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
  <Text
    style={
      {
        "color": "#145DA0",
        "fontFamily": "Roboto-Bold",
        "fontSize": 14,
        "marginBottom": 16,
        "textAlign": "center",
      }
    }
  >
    Bem vindo, 
    Test User
  </Text>
  <RCTScrollView
    ListEmptyComponent={
      <Styled(Text)>
        No brands found
      </Styled(Text)>
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <Text
        onLayout={[Function]}
        style={
          {
            "color": "#c4c4c4",
            "fontSize": 16,
            "marginTop": 16,
            "textAlign": "center",
          }
        }
      >
        No brands found
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
